File: lingopulse/ring_buffer.py
import fcntl
import json
import pathlib
from typing import Callable


class RingBuffer:
    def __init__(self, path: pathlib.Path, size: int = 5):
        self._path = path
        self._size = size

    def _read(self, f) -> list[dict]:
        f.seek(0)
        raw = f.read()
        if not raw:
            return []
        return json.loads(raw).get("entries", [])

    def _write(self, f, entries: list[dict]) -> None:
        f.seek(0)
        f.truncate()
        f.write(json.dumps({"entries": entries}, ensure_ascii=False))
        f.flush()

    def append(self, entry: dict) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        with self._path.open("a+", encoding="utf-8") as f:
            fcntl.flock(f, fcntl.LOCK_EX)
            entries = self._read(f)
            entries.append(entry)
            if len(entries) > self._size:
                entries = entries[-self._size:]
            self._write(f, entries)

    def pop_latest(self) -> dict | None:
        if not self._path.exists():
            return None
        with self._path.open("r+", encoding="utf-8") as f:
            fcntl.flock(f, fcntl.LOCK_EX)
            entries = self._read(f)
            if not entries:
                return None
            entry = entries.pop()
            self._write(f, entries)
            return entry

    def list_all(self) -> list[dict]:
        if not self._path.exists():
            return []
        with self._path.open("r", encoding="utf-8") as f:
            fcntl.flock(f, fcntl.LOCK_SH)
            entries = self._read(f)
        return list(reversed(entries))

    def find_matching(self, predicate: Callable[[dict], bool]) -> dict | None:
        for entry in self.list_all():
            if predicate(entry):
                return entry
        return None

File: app/Sources/LingoPulseApp/Services/RingBuffer.swift
import Foundation
import Darwin

actor RingBuffer {
    private let fileURL: URL
    private let size: Int

    init(fileURL: URL = RingBuffer.defaultURL(), size: Int = 5) {
        self.fileURL = fileURL
        self.size = size
    }

    static func defaultURL() -> URL {
        let home = FileManager.default.homeDirectoryForCurrentUser
        return home.appendingPathComponent(".cache/lingopulse/undo_ring.json")
    }

    func append(_ entry: [String: Any]) throws {
        let dir = fileURL.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        if !FileManager.default.fileExists(atPath: fileURL.path) {
            FileManager.default.createFile(atPath: fileURL.path, contents: nil)
        }

        let handle = try FileHandle(forUpdating: fileURL)
        defer {
            flock(handle.fileDescriptor, LOCK_UN)
            handle.closeFile()
        }
        flock(handle.fileDescriptor, LOCK_EX)

        var entries = readEntries(handle: handle)
        entries.append(entry)
        if entries.count > size {
            entries = Array(entries.suffix(size))
        }
        try writeEntries(entries, handle: handle)
    }

    func popLatest() throws -> [String: Any]? {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return nil }

        let handle = try FileHandle(forUpdating: fileURL)
        defer {
            flock(handle.fileDescriptor, LOCK_UN)
            handle.closeFile()
        }
        flock(handle.fileDescriptor, LOCK_EX)

        var entries = readEntries(handle: handle)
        guard !entries.isEmpty else { return nil }
        let last = entries.removeLast()
        try writeEntries(entries, handle: handle)
        return last
    }

    func listAll() throws -> [[String: Any]] {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return [] }

        let handle = try FileHandle(forReadingFrom: fileURL)
        defer {
            flock(handle.fileDescriptor, LOCK_UN)
            handle.closeFile()
        }
        flock(handle.fileDescriptor, LOCK_SH)

        return readEntries(handle: handle).reversed()
    }

    func findMatching(_ predicate: ([String: Any]) -> Bool) throws -> [String: Any]? {
        try listAll().first(where: predicate)
    }

    private func readEntries(handle: FileHandle) -> [[String: Any]] {
        handle.seek(toFileOffset: 0)
        let data = handle.readDataToEndOfFile()
        guard !data.isEmpty,
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let entries = obj["entries"] as? [[String: Any]]
        else { return [] }
        return entries
    }

    private func writeEntries(_ entries: [[String: Any]], handle: FileHandle) throws {
        let data = try JSONSerialization.data(withJSONObject: ["entries": entries], options: [])
        handle.seek(toFileOffset: 0)
        handle.truncateFile(atOffset: 0)
        handle.write(data)
    }
}
